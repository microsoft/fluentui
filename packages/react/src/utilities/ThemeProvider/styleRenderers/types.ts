import { IFontFace, IKeyframes } from '@fluentui/merge-styles';

/* eslint-disable @typescript-eslint/naming-convention */

type StyleRendererOptions = {
  rtl?: boolean;
  targetWindow: Window | undefined;
};

export interface StyleRenderer {
  /**
   * Expected to initialize or re-initialize the renderer. Primarily for testing purposes.
   */
  reset: () => void;

  /**
   * Returns a unique id for the renderer; used as part of the cache key when determining if new
   * styles need to be rendered.
   */
  getId: () => number;

  /**
   * Renders a stylesheet and returns the map of key to class name.
   */
  renderStyles: <TRuleSet>(ruleSet: TRuleSet, options: StyleRendererOptions) => { [key in keyof TRuleSet]: string };

  /**
   * Renders keyframes and returns the keyframe name.
   */
  renderKeyframes: (keyframes: IKeyframes, options: StyleRendererOptions) => string;

  /**
   * Renders fontfaces.
   */
  renderFontFace: (fontFace: IFontFace, options: StyleRendererOptions) => void;
}
