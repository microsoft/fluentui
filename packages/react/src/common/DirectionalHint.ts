export const DirectionalHint = {
  /**
   * Appear above the target element, with the left edges of the callout and target aligning.
   */
  topLeftEdge: 0 as 0,

  /**
   * Appear above the target element, with the centers of the callout and target aligning.
   */
  topCenter: 1 as 1,

  /**
   * Appear above the target element, with the right edges of the callout and target aligning.
   */
  topRightEdge: 2 as 2,

  /**
   * Appear above the target element, aligning with the target element such that the callout tends toward
   * the center of the screen.
   */
  topAutoEdge: 3 as 3,

  /**
   * Appear below the target element, with the left edges of the callout and target aligning.
   */
  bottomLeftEdge: 4 as 4,

  /**
   * Appear below the target element, with the centers of the callout and target aligning.
   */
  bottomCenter: 5 as 5,

  /**
   * Appear below the target element, with the right edges of the callout and target aligning.
   */
  bottomRightEdge: 6 as 6,

  /**
   * Appear below the target element, aligning with the target element such that the callout tends toward
   * the center of the screen.
   */
  bottomAutoEdge: 7 as 7,

  /**
   * Appear to the left of the target element, with the top edges of the callout and target aligning.
   */
  leftTopEdge: 8 as 8,

  /**
   * Appear to the left of the target element, with the centers of the callout and target aligning.
   */
  leftCenter: 9 as 9,

  /**
   * Appear to the left of the target element, with the bottom edges of the callout and target aligning.
   */
  leftBottomEdge: 10 as 10,

  /**
   * Appear to the right of the target element, with the top edges of the callout and target aligning.
   */
  rightTopEdge: 11 as 11,

  /**
   * Appear to the right of the target element, with the centers of the callout and target aligning.
   */
  rightCenter: 12 as 12,

  /**
   * Appear to the right of the target element, with the bottom edges of the callout and target aligning.
   */
  rightBottomEdge: 13 as 13,
};

export type DirectionalHint = typeof DirectionalHint[keyof typeof DirectionalHint];
