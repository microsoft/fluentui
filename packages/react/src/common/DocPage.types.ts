import { IStyleFunctionOrObject } from '../Utilities';
import { ITheme, IStyle } from '../Styling';

export interface IExample {
  /** Title of the example */
  title: string;

  /** Raw source code of the example */
  code: string;

  /** Working example of the example */
  view: JSX.Element;

  isScrollable?: boolean;

  /** JS String for codepen of the example */
  codepenJS?: string;

  /** Custom styles. Partial version of `IExampleCardProps['styles']`. */
  styles?: IStyleFunctionOrObject<{ theme?: ITheme }, { root: IStyle }>;
}

export interface IDocPageProps {
  /** Title that goes into the header */
  title: string;

  /** Name of the component being documented */
  componentName: string;

  /** URL of the checked in component, should be somewhere on github.com */
  componentUrl: string;

  /** Knobs that applies to all the examples */
  exampleKnobs?: JSX.Element;

  /** Array of examples, displayed in the order defined */
  examples?: IExample[];

  /** Properties table(s) as markdown string */
  propertiesTablesSources?: string[];

  /** Overview of the component as markdown string */
  overview?: string;

  /** Accessibility of the component as markdown string */
  accessibility?: string;

  /** DO's blurb as markdown string */
  dos?: string;

  /** DON'Ts blurb as markdown string */
  donts?: string;

  /** Best practice as markdown string */
  bestPractices?: string;

  /** Feedback section includes link to new issue page and displays Github issues */
  isFeedbackVisible?: boolean;

  /** Passed through header visibility flag from the demo component page component */
  isHeaderVisible: boolean;

  /** If true, the component accepts all native props from elements specified in `nativePropsElement` */
  allowNativeProps?: boolean;

  /** Override component name to use in the native props message */
  allowNativePropsForComponentName?: string;

  /**
   * Element(s) whose native props this component accepts (default div).
   * Only relevant if `allowNativeProps` is true.
   */
  nativePropsElement?: string | string[];

  /**
   * Related link
   * @deprecated No longer shown on ComponentPage
   */
  related?: JSX.Element;

  /** Pass through other sections for ComponentPage */
  otherSections?: {
    title: string;
    section: JSX.Element;
  }[];

  /**
   * JSON to populate the api reference tables
   */
  jsonDocs?: IPageJson;
}

/**
 * Used to keep track of where an API reference page will live on the site.
 */
export type PageKind = 'References' | 'Components';

/**
 * Text excerpt token that is part of a type definition or extends block and may have a link
 * to another doc page. For API reference tables.
 */
export interface ILinkToken {
  text: string;
  /** If this token is a link, name of the doc page it points to */
  linkedPage?: string;
  /** If this token is a link, group/category of the doc page it points to */
  linkedPageGroup?: string;
}

/**
 * Generic row for API reference tables.
 * It can represent a member (property or method) of an interface or class.
 */
export interface ITableRowJson {
  name: string;
  kind?: 'method' | 'property';
  /**
   * The row's type translated to an array of text elements and links to other types.
   * For example, `Readonly<IFoo>` would translate to:
   * `[{ text: 'Readonly<' }, { text: 'IFoo', hyperlinkedPage: '(page name)', pageKind: '(kind)' }, { text: '>' }]`
   */
  typeTokens: ILinkToken[];
  defaultValue?: string;
  description: string;
  deprecated: boolean;
  deprecatedMessage?: string;
  required?: boolean;
}

/**
 * Enum member row for API reference tables.
 */
export type IEnumTableRowJson = Omit<ITableRowJson, 'kind' | 'typeTokens' | 'defaultValue' | 'required'> & {
  value: string;
};

export type ApiKind = 'interface' | 'enum' | 'class' | 'typeAlias';

/**
 * Info for a table representing a top-level API item: interface, enum, class, or type alias.
 */
export interface ITableJson {
  kind: ApiKind;
  name: string;
  /**
   * Any types the item extends, translated to an array of text elements and links to other types.
   * For classes and interfaces only.
   */
  extendsTokens: ILinkToken[];
  description: string;
  members?: ITableRowJson[] | IEnumTableRowJson[];
  deprecated?: boolean;
  deprecatedMessage?: string;
}

/**
 * Structure of the page.json files
 */
export interface IPageJson {
  tables: ITableJson[];
  name: string;
  group?: string;
}
