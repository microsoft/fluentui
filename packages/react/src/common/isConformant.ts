import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant<TProps = {}>(
  testInfo: Omit<IsConformantOptions<TProps>, 'componentPath'> & { componentPath?: string },
) {
  const defaultOptions: Partial<IsConformantOptions<TProps>> = {
    disabledTests: [
      'has-docblock',
      'kebab-aria-attributes',
      'is-static-property-of-parent',
      // Will enable with appropriate overrides separately
      'consistent-callback-names',
    ],
    skipAsPropTests: true,
    componentPath: module!.parent!.filename.replace('.test', ''),
  };

  baseIsConformant(defaultOptions, testInfo);
}
