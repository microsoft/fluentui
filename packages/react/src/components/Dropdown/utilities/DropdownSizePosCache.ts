import { IDropdownOption, DropdownMenuItemType } from '../Dropdown.types';

/**
 * A utility class to cache size and position in cache.
 *
 * Dropdown options has non-selectable display types. It is therefore not cheap to determine
 * the total number of actual selectable options as well as the position an option is in the
 * list of options - O(n) cost for each lookup.
 *
 * Given that we potentially have to make this determination on every single render pass, this
 * cache should provide a little bit of relief.
 */
export class DropdownSizePosCache {
  private _cachedOptions: IDropdownOption[];
  private _displayOnlyOptionsCache: number[];
  private _size = 0;

  /**
   * Invalidates the cache and recalculate the size of selectable options.
   */
  public updateOptions(options: IDropdownOption[]) {
    const displayOnlyOptionsCache = [];
    let size = 0;
    for (let i = 0; i < options.length; i++) {
      if (options[i].itemType === DropdownMenuItemType.Divider || options[i].itemType === DropdownMenuItemType.Header) {
        displayOnlyOptionsCache.push(i);
      } else if (!options[i].hidden) {
        size++;
      }
    }

    this._size = size;
    this._displayOnlyOptionsCache = displayOnlyOptionsCache;
    this._cachedOptions = [...options];
  }

  /**
   * The size of all the selectable options.
   */
  public get optionSetSize(): number {
    return this._size;
  }

  /**
   * The chached options array.
   */
  public get cachedOptions(): IDropdownOption[] {
    return this._cachedOptions;
  }

  /**
   * Returns the position of this option element relative to the full set of selectable option elements.
   * Note: the first selectable element is position 1 in the set.
   * @param index The raw index of the option element.
   */
  public positionInSet(index: number | undefined): number | undefined {
    if (index === undefined) {
      return undefined;
    }

    // we could possibly memoize this too but this should be good enough, most of the time (the expectation is that
    // when you have a lot of options, the selectable options will heavily dominate over the non-selectable options.
    let offset = 0;
    while (index > this._displayOnlyOptionsCache[offset]) {
      offset++;
    }

    if (this._displayOnlyOptionsCache[offset] === index) {
      throw new Error(`Unexpected: Option at index ${index} is not a selectable element.`);
    }

    return index - offset + 1;
  }
}
