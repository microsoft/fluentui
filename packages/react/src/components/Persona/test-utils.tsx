import * as React from 'react';
import type { IRenderFunction } from '../../Utilities';
import type { IPersonaProps, IPersonaSharedProps } from './index';

/**
 * function to override the default onRender callbacks
 */
export const wrapPersona = (
  example: IPersonaSharedProps,
  shouldWrapPersonaCoin: boolean = false,
): ((coinProps: IPersonaProps, defaultRenderer: IRenderFunction<IPersonaProps>) => JSX.Element | null) => {
  return (coinProps, defaultCoinRenderer): JSX.Element | null => {
    return shouldWrapPersonaCoin ? (
      <span id="persona-coin-container">{defaultCoinRenderer(coinProps)}</span>
    ) : (
      defaultCoinRenderer(coinProps)
    );
  };
};
