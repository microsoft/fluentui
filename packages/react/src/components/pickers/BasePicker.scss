@import '~@fluentui/common-styles/src/common';

.pickerText {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  box-sizing: border-box;
  border: 1px solid $ms-color-neutralTertiary;
  min-width: 180px;
  min-height: 30px;
  &:hover {
    border-color: $inputBorderHoveredColor;
  }

  &.inputFocused {
    position: relative;
    border-color: $inputFocusBorderAltColor;
    &:after {
      pointer-events: none;
      content: '';
      position: absolute;
      left: -1px;
      top: -1px;
      bottom: -1px;
      right: -1px;
      border: 2px solid $inputFocusBorderAltColor;
    }
  }

  &.inputDisabled {
    @include high-contrast {
      position: relative;
      border-color: GrayText;
      &:after {
        pointer-events: none;
        content: '';
        position: absolute;
        left: 0px;
        top: 0px;
        bottom: 0px;
        right: 0px;
        background-color: Window;
      }
    }
  }
}

.pickerInput {
  height: 34px;
  border: none;
  flex-grow: 1;
  outline: none;
  padding: 0 6px 0px;
  align-self: flex-end;
}

.pickerItems {
  display: flex;
  flex-wrap: wrap;
  max-width: 100%;
}

.screenReaderOnly {
  @include ms-screenReaderOnly;
}
