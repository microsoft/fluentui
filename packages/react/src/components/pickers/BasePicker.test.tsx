import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
import * as renderer from 'react-test-renderer';

import { IBasePickerProps, IBasePicker, ValidationState } from './BasePicker.types';
import { BasePicker } from './BasePicker';
import { IPickerItemProps } from './PickerItem.types';
import { resetIds, KeyCodes } from '@fluentui/utilities';
import { isConformant } from '../../common/isConformant';

function onResolveSuggestions(text: string): ISimple[] {
  return [
    'black',
    'blue',
    'brown',
    'cyan',
    'green',
    'magenta',
    'mauve',
    'orange',
    'pink',
    'purple',
    'red',
    'rose',
    'violet',
    'white',
    'yellow',
  ]
    .filter(tag => tag.toLowerCase().indexOf(text.toLowerCase()) === 0)
    .map(item => ({ key: item, name: item }));
}

export interface ISimple {
  key: string;
  name: string;
}

const basicRenderer = (props: IPickerItemProps<ISimple>) => {
  return <div> {props.item.name} </div>;
};

const basicSuggestionRenderer = (props: ISimple) => {
  return <div> {props.name} </div>;
};

const getSuggestions = (root: HTMLElement | Document) => {
  return document.querySelector<HTMLElement>('.ms-Suggestions');
};

describe('BasePicker', () => {
  let root: HTMLDivElement;
  beforeEach(() => {
    root = document.createElement('div');
    resetIds();
  });

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(root);
    document.body.textContent = '';

    // reset any jest timers
    if ((setTimeout as any).mock) {
      jest.runOnlyPendingTimers();
      jest.useRealTimers();
    }
  });

  const BasePickerWithType = BasePicker as new (props: IBasePickerProps<ISimple>) => BasePicker<
    ISimple,
    IBasePickerProps<ISimple>
  >;
  const onRenderItem = (props: IPickerItemProps<ISimple>): JSX.Element => (
    <div key={props.item.name}>{basicRenderer(props)}</div>
  );

  const runAllTimers = () =>
    ReactTestUtils.act(() => {
      jest.runAllTimers();
    });

  it('renders correctly', () => {
    const component = renderer.create(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with inputProps supply classnames correctly', () => {
    const component = renderer.create(
      <BasePickerWithType
        inputProps={{
          placeholder: 'Bitte einen Benutzer angeben...',
          id: 'pckSelectedUser',
          className: 'testclass ',
        }}
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  isConformant({
    Component: BasePicker,
    displayName: 'BasePicker',
    // Problem: Ref doesn't match DOM node and returns null.
    // Solution: Ensure ref is passed correctly to the root element.
    disabledTests: ['component-has-root-ref', 'component-handles-ref', 'has-top-level-file'],
  });

  it('renders inline callout', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);
    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
        pickerCalloutProps={{ doNotLayer: true, id: 'test' }}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'b';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    const calloutParent = document.getElementById('test')?.closest('.ms-BasePicker');
    expect(calloutParent).toBeTruthy();
  });

  it('can provide custom renderers', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();
    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'bl';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();

    const suggestionOptions = document.querySelectorAll('.ms-Suggestions-itemButton');
    expect(suggestionOptions.length).toEqual(2);
    ReactTestUtils.Simulate.click(suggestionOptions[0]);

    const currentPicker = picker.current!.items;
    expect(currentPicker).toHaveLength(1);
    expect(currentPicker![0].name).toEqual('black');
  });

  it('can select generic items.', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();
    const onValidateInput = () => {
      return ValidationState.valid;
    };

    const createGenericItem = (str: string): ISimple => {
      return {
        key: str,
        name: str,
      };
    };

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
        createGenericItem={createGenericItem}
        onValidateInput={onValidateInput}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'asdff';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();

    const suggestionOptions = document.querySelectorAll('.ms-Suggestions-itemButton');
    expect(suggestionOptions.length).toEqual(0);
    ReactTestUtils.Simulate.keyDown(input, { which: KeyCodes.enter });

    const currentPicker = picker.current!.items;
    expect(currentPicker).toHaveLength(1);
    expect(currentPicker![0].name).toEqual('asdff');
  });

  it('has force suggestions button', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();
    const onValidateInput = () => {
      return ValidationState.valid;
    };

    const createGenericItem = (str: string): ISimple => {
      return {
        key: str,
        name: str,
      };
    };

    const showForceSuggestionsText = () => true;

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
        createGenericItem={createGenericItem}
        onValidateInput={onValidateInput}
        pickerSuggestionsProps={{
          showForceResolve: showForceSuggestionsText,
          forceResolveText: 'Force',
        }}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'asdff';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();

    const forceButton = document.querySelectorAll('[data-automationid=sug-forceResolve]');
    expect(forceButton.length).toEqual(1);
    ReactTestUtils.Simulate.click(forceButton[0]);

    const currentPicker = picker.current!.items;
    expect(currentPicker).toHaveLength(1);
    expect(currentPicker![0].name).toEqual('asdff');
  });

  it('will not render input when items reach itemLimit', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        itemLimit={1}
        componentRef={picker}
      />,
      root,
    );

    let input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'bl';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    const suggestionOptions = document.querySelectorAll('.ms-Suggestions-itemButton');
    ReactTestUtils.Simulate.click(suggestionOptions[0]);

    const currentPicker = picker.current!.items;
    expect(currentPicker).toHaveLength(1);

    input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    expect(input).toBeNull();
  });

  it('will still render with itemLimit set to 0', () => {
    document.body.appendChild(root);

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        itemLimit={0}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    expect(input).toBeNull();
  });

  it('can be set with selectedItems and a lower itemLimit', () => {
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <BasePickerWithType
        selectedItems={[
          { key: '1', name: 'blue' },
          { key: '2', name: 'black' },
        ]}
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        itemLimit={0}
        componentRef={picker}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    expect(input).toBeNull();

    const currentPicker = picker.current!.items;
    expect(currentPicker).toHaveLength(2);
  });

  it('can render MRU when input is focused', () => {
    const resolveSuggestions = () => onResolveSuggestions('');
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onEmptyInputFocus={resolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();

    expect(getSuggestions(document)).toBeTruthy();

    const suggestionOptions = document.querySelectorAll('.ms-Suggestions-itemButton');
    expect(suggestionOptions.length).toEqual(15);

    const currentPicker = picker.current!.items;
    expect(currentPicker).toHaveLength(0);
  });

  it('Closes menu when escape is pressed', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'asdff';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();

    ReactTestUtils.Simulate.keyDown(input, { which: 27 });

    expect(getSuggestions(document)).toBeFalsy();
  });

  it('Opens menu on click when suggestions have been dismissed', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'asdff';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();

    ReactTestUtils.Simulate.keyDown(input, { which: 27 });

    expect(getSuggestions(document)).toBeFalsy();
    ReactTestUtils.Simulate.click(input, { button: 0 });

    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();
  });

  it('Opens menu when input refocused after search has happened', () => {
    expect(getSuggestions(document)).toBeFalsy();
    jest.useFakeTimers();
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <div>
        <BasePickerWithType
          onResolveSuggestions={onResolveSuggestions}
          onRenderItem={onRenderItem}
          onRenderSuggestionsItem={basicSuggestionRenderer}
          componentRef={picker}
        />
        <button id="toFocus">focus me</button>
      </div>,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'bl';
    ReactTestUtils.Simulate.input(input);
    // For some reason with act() this has to be run twice to make the callout dismiss callback
    // actually be called?
    runAllTimers();
    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();

    (document.querySelector('#toFocus') as any).focus();

    // Implicit test to ensure suggestions are dismissed when focus lost
    expect(getSuggestions(document)).toBeFalsy();

    runAllTimers();
    input.focus();
    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();
  });

  it('Opens calls onResolveSuggestions if it currently doesnt have suggestions', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    let count = 0;
    const resolveCounter = (val: string) => {
      count++;
      return onResolveSuggestions(val);
    };
    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={resolveCounter}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
        inputProps={{ defaultVisibleValue: 'bl' }}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    runAllTimers();

    expect(count).toEqual(1);

    expect(getSuggestions(document)).toBeTruthy();
  });

  it('navigates to search for more button', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();
    const onValidateInput = () => {
      return ValidationState.valid;
    };

    const createGenericItem = (str: string): ISimple => {
      return {
        key: str,
        name: str,
      };
    };

    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        componentRef={picker}
        createGenericItem={createGenericItem}
        onValidateInput={onValidateInput}
        pickerSuggestionsProps={{
          searchForMoreText: 'More Options',
        }}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'b';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    expect(getSuggestions(document)).toBeTruthy();

    const moreButton = document.querySelector('[data-automationid=sug-searchForMore]') as HTMLElement;
    expect(moreButton).toBeTruthy();
    ReactTestUtils.Simulate.keyDown(input, { which: KeyCodes.up });

    expect(moreButton.id).toEqual('sug-selectedAction');
    expect(input.getAttribute('aria-activedescendant')).toEqual('sug-selectedAction');
  });

  it('focuses the input when the focus method is called', () => {
    document.body.appendChild(root);

    const picker = React.createRef<IBasePicker<ISimple>>();

    ReactDOM.render(
      <BasePickerWithType
        componentRef={picker}
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    picker.current?.focus();

    expect(document.activeElement).toBe(input);
  });

  it('focuses the last selected item after removing input', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const onRenderFocusableItem = (props: IPickerItemProps<ISimple>): JSX.Element => (
      <button key={props.item.name} data-selection-index={props.index}>
        {basicRenderer(props)}
      </button>
    );
    ReactDOM.render(
      <BasePickerWithType
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderFocusableItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
        itemLimit={1}
      />,
      root,
    );

    const input = document.querySelector('.ms-BasePicker-input') as HTMLInputElement;
    input.focus();
    input.value = 'bl';
    ReactTestUtils.Simulate.input(input);
    runAllTimers();

    const suggestionOptions = document.querySelectorAll('.ms-Suggestions-itemButton');
    ReactTestUtils.Simulate.click(suggestionOptions[0]);

    const selectedItem = document.querySelector('button[data-selection-index]');

    expect(document.activeElement).toBe(selectedItem);
  });

  it('focuses the next selected item after removing a selection', () => {
    jest.useFakeTimers();
    document.body.appendChild(root);

    const onRenderFocusableItem = (props: IPickerItemProps<ISimple>): JSX.Element => {
      return (
        <button key={props.item.name} onClick={props.onRemoveItem} data-selection-index={props.index}>
          {basicRenderer(props)}
        </button>
      );
    };
    ReactDOM.render(
      <BasePickerWithType
        defaultSelectedItems={[
          { key: '1', name: 'blue' },
          { key: '2', name: 'black' },
        ]}
        onResolveSuggestions={onResolveSuggestions}
        onRenderItem={onRenderFocusableItem}
        onRenderSuggestionsItem={basicSuggestionRenderer}
      />,
      root,
    );

    const selectedEls = document.querySelectorAll('button[data-selection-index]');
    (selectedEls[0] as HTMLButtonElement).focus();
    ReactTestUtils.Simulate.click(selectedEls[0]);

    jest.runAllTimers();

    expect(document.activeElement).toBe(selectedEls[1]);
  });
});
