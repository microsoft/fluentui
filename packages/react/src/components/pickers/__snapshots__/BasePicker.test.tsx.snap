// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasePicker renders correctly 1`] = `
<div
  className="ms-BasePicker"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
>
  <span
    hidden={true}
    id="selected-items-id__0-label"
  />
  <div
    className="ms-SelectionZone"
    onClick={[Function]}
    onContextMenu={[Function]}
    onDoubleClick={[Function]}
    onFocusCapture={[Function]}
    onKeyDown={[Function]}
    onKeyDownCapture={[Function]}
    onMouseDown={[Function]}
    onMouseDownCapture={[Function]}
    role="presentation"
  >
    <div
      className="ms-BasePicker-text"
    >
      <input
        aria-autocomplete="both"
        aria-controls=""
        aria-expanded={false}
        aria-haspopup="listbox"
        autoCapitalize="off"
        autoComplete="off"
        className="ms-BasePicker-input"
        data-lpignore={true}
        id="combobox-id__0"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onCompositionEnd={[Function]}
        onCompositionStart={[Function]}
        onCompositionUpdate={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        role="combobox"
        spellCheck={false}
        style={
          Object {
            "fontFamily": "inherit",
          }
        }
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`BasePicker renders with inputProps supply classnames correctly 1`] = `
<div
  className="ms-BasePicker"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
>
  <span
    hidden={true}
    id="selected-items-id__0-label"
  />
  <div
    className="ms-SelectionZone"
    onClick={[Function]}
    onContextMenu={[Function]}
    onDoubleClick={[Function]}
    onFocusCapture={[Function]}
    onKeyDown={[Function]}
    onKeyDownCapture={[Function]}
    onMouseDown={[Function]}
    onMouseDownCapture={[Function]}
    role="presentation"
  >
    <div
      className="ms-BasePicker-text"
    >
      <input
        aria-autocomplete="both"
        aria-controls=""
        aria-expanded={false}
        aria-haspopup="listbox"
        autoCapitalize="off"
        autoComplete="off"
        className="ms-BasePicker-input testclass "
        data-lpignore={true}
        id="pckSelectedUser"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onCompositionEnd={[Function]}
        onCompositionStart={[Function]}
        onCompositionUpdate={[Function]}
        onFocus={[Function]}
        onInput={[Function]}
        onKeyDown={[Function]}
        placeholder="Bitte einen Benutzer angeben..."
        role="combobox"
        spellCheck={false}
        style={
          Object {
            "fontFamily": "inherit",
          }
        }
        value=""
      />
    </div>
  </div>
</div>
`;
